### ISSUE TITLE
[ACI post-project connection] Implement consolidated logic for “Cypress-triggered events” like displaying the CI prompt, smart banners, and post-login messages.

### ISSUE THREAD

**Issue by marktnoonan** [cypress-io/cypress#23765]
**Context**
This [Whimsical diagram](https://whimsical.com/get-to-first-run-TgixRWciqE1chPGDCHk5b2)(only visible to internal Cy employees) gathers together the logic that has evolved to manage smart banners, the old CI prompt, LoginConnect modal state, as well as the state of various buttons in the Top Nav, Specs List Page, Settings Page that allow users to log in and/or connect a project to he Dashboard.

The diagram represents User-Triggered Actions with solid arrows between states, and Cypress-Triggered Actions with dotted arrows. Cypress-triggered actions like displaying a notification are intended to only happen when certain specific rules have been met, and also shouldn't happen close together, so as to avoid becoming annoying, and increase the chance that they are useful.

Currently the logic around these things has been implemented similarly in more than one place, and it's time to consolidate all of it so that we can easily add new features that depend on this logic.

**Requested Change**

A solution is needed that gathers all of the logic from that Whimsical into a single system whose overall purpose is to

1. determine if Cypress has permission to do something, and 
2. provide a reactive single source of truth to define which of the small number of possible states a user is currently in. 

For example:

SpecsListBanners.vue currently uses a watcher to set various flags as the user logs in and a project gets setup, so it can determine which banner variant should be shown. It includes logic for "whether the user state is correct" (has a project? etc) to show a certain banner, and "whether Cypress is allowed" to show the banner (has more than 4 days of Cypress use?). Using the new solution, when SpecsListBanner mounts it would check which of the N possible states a user is in (which is provided from a store or similar) and if that state maps to a variant of the banner, it would call a function that checks if it has permission to open. After opening and after being dismissed, it would make the appropriate calls to this helper to record the updated date and keep everything in sync.

The "what state is the user in" question is partially solved by the `login-connect-store` introduced in #23762 and the related `CloudViewerAndProject` component that provides state via a scoped slot. But these are essentially stepping stones to a more holistic system. That branch has not been merged. It could be a starting point for this work, or just a reference point to start fresh.

The "whether Cypress is allowed" to do something question is not centrally managed at the moment. It is about to get a little more complex as now there is a 1-day delay being introduced.

**Rationale for doing this**
One reason to gather this logic centrally is to have a smoother developer experience using these kinds of values and figuring out if we can display something to the user in an increasingly complicated set of circumstances. It should be clear how to quickly "add a button" anywhere and tell it how to behave in the context of the LoginConnect state, without having to recreate the logic to derive that state. This should make us more nimble with future requests.

Another developer reason to extract this stuff is that it has nothing to do with Vue, this business logic can and should be unit tested on its own and independent of a UI framework

There's a third reason to pause and do this now: to get aligned with Product in terms of how we are naming the states users can be in, and how we define the rules for Cypress-triggered actions. The less translation we need when we talk about these things and implementing them, the better. Chris France is already using state charts to ferret out all the allowed states of these flows for figma and prototype reasons. While I'm not suggesting we get a full state machine pattern going here, we should be thinking in that spirit, and especially we should go out of our way to reject impossible states, which might lead to showing the banner and the CI prompt at the same time, or something like that. I'd expect long term this kind of thing might evolve towards a state machine.

For now a clear and human readable expression of the rules, plus some validation, is likely good enough (see suggested implementation below).

**Acceptance Criteria**

- user state like `const isProjectConnected = !!currentProject?.projectId && currentProject.cloudProject?.__typename === 'CloudProject'` is computed in one place and accessible to components that need it
- cypress-triggered actions check if they have permission to happen, and record when they have happened
- business logic related to the states and transitions in Whimsical is encapsulated in plain TS helpers that are unit tested, and the unit tests describe all of the expected scenarios
- logic from the Whimsical (like a 1-day delay) is implemented in the helper code and tested, but not wired up in the app yet, let's merge the helper, then do the wiring in #23768, #23767, and #23766.

**Suggested implementation** (for estimation purposes)

This could be done in many ways, I don't think this ticket needs to be too prescriptive as to the implementation. Though it could be broken down further into multiple steps. 

I do have an pseudocode example of how the code that manages this in a common language with Product might look, which would let us easily change rules or add new rules. This of course implies a set of functions and validators that aren't shown that would actually parse objects like this and apply the rules.

<details>
<summary>
Psuedocode for how we might manage the rules in a human readable way
</summary>

```js
const userStates = {  
// this names the 4 states defined in whimsical and expresses the rules in a human readable way
// a validator function should be used, and throw if 2 states have the same set of conditions
// we can only be in one of these at a time, ever
  loggedOut: {
    matches: ['notLoggedIn']
  },
  needsOrg: {
    matches: ['isLoggedIn', 'hasNoOrgs']
   }, 
  needsConnectedProject: {
    matches: [
      'isLoggedIn', 
      'hasAtLeastOneOrg', 
      'hasNoConnectedProject'
    ]
   }, 
  needsRecordedRun : {
    matches: [
      'isLoggedIn', 
      'hasAtLeastOneOrg',
      'hasConnectedProject', 
      'hasNoRecordedRuns', 
      minSpecsInCurrentTestingType(4)
    ]
   }  
}

const features = {
  smartBanner: { // feature - where we are in the app
    cypressTriggerRules: [ // conditions that must be true for Cypress to trigger this feature to open
      minDaysSince('firstOpen', 4), 
      minDaysSince('docsAutoOpen', 1),
      minDaysSince('connectProject', 1),
      minDaysSince('recordRunShown', 1),
      isInState(['loggedOut', 'needsOrg', 'needsConnectedProject', 'needsRecordedRun'])
      // needs a validator that the rules do not conflict
      // or can nest rules in paths to prevent impossible states
    ],
    content: { // content - customized per cohort - can live in existing i18n json
      loggedOut: { // user state - derived from gql
          a: 'Log in',
          b: 'Log in, please',
          default: 'Log in' // no cohort
      },
      needsOrg: {
          a: 'Do an org now',
          b: 'Make an org pls',
          default: 'Make org'
      }
    }
  },
  loginConnectModal: {
    // ...
  }
}

```
</details>

**Comment by baus** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Component-Testing-606e43bbf1b3b20013c142c1/issues/cypress-io/cypress/23765?planning-poker) with Zenhub @amehta265 @astone123 @lmiller1990 @marktnoonan @mike-plummer @rockindahizzy @warrensplayer @ZachJW34

**Comment by lmiller1990** [comment]
TL;DR you hit the nail on the head at the end: we need a (kind of) state machine to manage complexity. I 100% agree this should be decoupled from the UI layer. I wonder if XState is a good fit (probably not? steep learning curve, complicated) but it has Vue bindings: https://xstate.js.org/docs/recipes/vue.html. If we end up re-inventing a state machine anyway, maybe it would be a good fit, but we'd probably need to rejig things a bit to make it fit.

I think this ticket has a lot of thinking and planning involved. My estimate reflects that - I expect the engineer to spend the majority of development time thinking/basic prototyping, as opposed to just going with the first thing they come up with. Looping someone in early to bounce ideas might be a good move, too.

**Comment by marktnoonan** [comment]
I like the idea of moving to XState for something like this, but agree it seems to involve some ramp up time and might be a distraction. The advantage is that it would quite directly align with what @christopherfrance is doing to define these states on the design side, so it's possible that we could synchronize the implementation with that process. But still, I think that requires some level of buy-in that we don't have right now, and would take some time to get to. It might be a good idea to look at the use of XState in the real world app, and consider a similar API so that switching to xstate would be a natural step in the future.

I'll make a video shortly about what I've done already and talk about some of the pros and cons I've found, I think it might be a good pattern for how we'll getting things in and out of vue. Essentially:

1. A "UserState" (or whatever, naming this is tough) component in frontend-shared that gets the required fields from GQL that can be used once at the root of launchpad and app
2. UserState.vue watches GQL and calls a function to deal with the updated state as users login, add tests, etc (and does not render in the Runner).
3. That function process the values and sets the resulting single, current derived state in a store, that's the first part of I/O between our new helper and Vue for state.
4. Components that need to use the state can look at the store, but for basic use in templates we can also have a wrapper component that provides the state via slot props if we want
5. Components that need to "ask permission" to display something like a banner, will call a function checking
	1. if the component itself is allowed to render at all? Catching rules _defined on the feature_ like "no banners for 24 hours after seeing the CI prompt"
	1. if it is, is there a variant for the current state? (eg smartBanner -> loggedOut -> yep we have something to show)
	1. If there's a variant for the current state, are we allowed to show that specific content (catching rules _defined on the content_ like "has it been seen and dismissed already, or is there another special rule beyond the users state in play").
	1. if there's no blocker, show the content
	1. once the content has been shown or dismissed, send the information  to the new helper, which will handle updating the saved state, which updates gql, which kicks of the evaluation of UserState again
	1. this is the 2nd part I/O between the helper and Vue.

The helper that manages the state and the rules based on these inputs could be replaced with a state machine at some point, and start out working more simply, as long as the API doesn't change.

**Comment by baus** [comment]
@warrensplayer This seems to be related to the idea of having a "In app messaging" framework. I'm afraid if we keep implementing logic ad-hoc that the complexity will get away from us

**Comment by marktnoonan** [comment]
Here's a rough loom showing, based on our current code, what logic would move into the new system. It feels more lightweight to me as an extension of the work that's already been done than starting from scratch: https://www.loom.com/share/f3c8e32352b24300a5b4aaaa67ed155b

**Comment by warrensplayer** [comment]
@baus Yes, it is time that we look at doing this work.  The work that Mark has started in the linked PR that he walks through in the Loom above implements some great concepts to use or work off of.
